plugins/inputs/beat/beat.go

```diff
@@ -0,0 +1,278 @@
+  # host_header = "logstash.example.com"
+
+  ## Timeout for HTTP requests
+  timeout = "5s"
+
+  ## Optional HTTP Basic Auth credentials
+  # username = "username"
+  # password = "pa$$word"
+
+  ## Optional TLS Config
+  # tls_ca = "/etc/telegraf/ca.pem"
+  # tls_cert = "/etc/telegraf/cert.pem"
+  # tls_key = "/etc/telegraf/key.pem"
+  ## Use TLS but skip chain & host verification
+  # insecure_skip_verify = false
+`
+
+const description = "Read metrics exposed by Beat"
+
+const suffixInfo = "/"
+const suffixStats = "/stats"
+
+type BeatInfo struct {
+	Beat     string `json:"beat"`
+	Hostname string `json:"hostname"`
+	Name     string `json:"name"`
+	UUID     string `json:"uuid"`
+	Version  string `json:"version"`
+}
+
+type BeatStats struct {
+	Beat     map[string]interface{} `json:"beat"`
+	FileBeat interface{}            `json:"filebeat"`
+	Libbeat  interface{}            `json:"libbeat"`
+	System   interface{}            `json:"system"`
+}
+
+type Beat struct {
+	URL string `toml:"url"`
+
+	CollectBeatStats     bool `toml:"collect_beat_stats"`
+	CollectLibbeatStats  bool `toml:"collect_libbeat_stats"`
+	CollectSystemStats   bool `toml:"collect_system_stats"`
+	CollectFilebeatStats bool `toml:"collect_filebeat_stats"`
+
+	Username   string            `toml:"username"`
+	Password   string            `toml:"password"`
+	Method     string            `toml:"method"`
+	Headers    map[string]string `toml:"headers"`
+	HostHeader string            `toml:"host_header"`
+	Timeout    internal.Duration `toml:"timeout"`
+
+	tls.ClientConfig
+	client *http.Client
+}
+
+func NewBeat() *Beat {
+	return &Beat{
+		URL:                  "http://127.0.0.1:5066",
+		CollectBeatStats:     true,
+		CollectLibbeatStats:  true,
+		CollectSystemStats:   true,
+		CollectFilebeatStats: true,
+		Method:               "GET",
+		Headers:              make(map[string]string),
+		HostHeader:           "",
+		Timeout:              internal.Duration{Duration: time.Second * 5},
+	}
+}
+
+func (beat *Beat) Description() string {
+	return description
+}
+
+func (beat *Beat) SampleConfig() string {
+	return sampleConfig
+}
+
+// createHttpClient create a clients to access API
+func (beat *Beat) createHttpClient() (*http.Client, error) {
+	tlsConfig, err := beat.ClientConfig.TLSConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsConfig,
+		},
+		Timeout: beat.Timeout.Duration,
+	}
+
+	return client, nil
+}
+
+// gatherJsonData query the data source and parse the response JSON
+func (beat *Beat) gatherJsonData(url string, value interface{}) error {
+
+	var method string
+	if beat.Method != "" {
+		method = beat.Method
+	} else {
+		method = "GET"
+	}
+
+	request, err := http.NewRequest(method, url, nil)
+	if err != nil {
+		return err
+	}
+
+	if (beat.Username != "") || (beat.Password != "") {
+		request.SetBasicAuth(beat.Username, beat.Password)
+	}
+	for header, value := range beat.Headers {
```

Comment:
The local variable `value` might shadow the function argument `value`. To be honest I'm also not sure that I like the passing the result `value` by reference instead of constructing it here and return it. Is there any reason for doing this?

Reply:
not sure if I understand the latter part of your comment, this bit is just applying the headers configuration. will rename the variables from the range to avoid them clashing with the value function parameter, LMK if there's other concerns.

Reply:
What I mean is instead of passing `value` to `gatherJsonData()` you could do `value := &BeatInfo{}` *in* that function and then return `value` to the caller.

Reply:
ahh yeah, that is a bit weird. I'm not the original author so I don't have the full context as far as history or intentions, but I'll fix this one.

Reply:
actually, I understand now. :) It's passed a `BeatInfo` and later a `BeatStats`, so it needs to accept an `interface{}`